pystac/__init__.py

```diff
@@ -71,7 +72,7 @@
 )
 
 
-def read_file(href: str) -> STACObject:
+def read_file(href: str) -> Union[STACObject, ItemCollection]:
```

Comment:
I'm not sure it makes sense to change these top level package read/write methods to include ItemCollection. It breaks user code that may rely on the STACObject type, forcing the type differentiation to happen by the caller. Also the last two parameters of `write_file` make it feel a bit shoe-horned. On the other hand, I can see people getting confused about  why `read_file` wouldn't work on an ItemCollection if they didn't know it wasn't a core stac object.

I think my preference would be to keep these methods working with core STAC types, and force users to treat ItemCollections as a separate concept, which makes more sense with the spec as it currently stands.

Reply:
Yeah, I agree with all of the points you make here, and I think my preference would be to not change the top-level functions as well. If that ends up being the decision, I will add more clear documentation to the `ItemCollection` class indicating that it is not a `STACObject` and cannot be read using those top-level methods.

@scottyhq I'm curious how much of a priority it is for you to be able to read Item Collections using `pystac.read_file` vs. `ItemCollection.from_file`.

Reply:
Thanks for all the work on this! Yes, I'm coming at this primarily from someone new to STAC, who is unfamiliar with spec details. The key expectation I have is that after searching a STAC API for data, and saving 'results.json' to work with later, there is an straightforward way to open that file and navigate it. For Python it seems like `pystac_client` takes care of the searching and `pystac` should care of the I/O and navigation of the results. I don't think people should have to understand the concepts of ItemCollections versus Collections, Core vs Not, for this fundamental workflow. 

So if `pystac.read_file` can't handle ItemCollections, and a separate `ItemCollection.from_file()` is the way forward (or pystac_client.read_file()`?), I think that just needs to be clearly documented. 

Also useful (and I think a non-breaking change) would then be for `pystac.read_file` to have error handing that can recognize the json is an ItemCollection and suggest the correct method to open it, rather than the current `KeyError: 'id'` ?

Reply:
Thanks @scottyhq, that all makes sense to me. I'll change those top-level functions back to their original signatures and make sure we have clear docs on how to work with ItemCollections. I'm pretty sure the issue with a `KeyError` being raised in `pystac.read_file` is fixed by #402, but I'll add a test to be sure.

@matthewhanson Looking back at the code example in the original issue it seems like the name of the [`ItemSearch.items_as_collection`](https://github.com/stac-utils/pystac-client/blob/9e66bb5b46951988f5be23a09579e512e0fd425f/pystac_client/item_search.py#L402) method might also be misleading. Maybe we should rename that to `items_as_item_collection` so that users don't think they are saving STAC Collection?

Reply:
Removed `ItemCollection` handling from top-level functions and added test that `pystac.read_file` raises a `STACTypeError` instead of the `KeyError` in 404bb99390086becce038b35d18a97a8e4d6160b

Reply:
@duckontheweb `ItemSearch.items_as_collection` was misleading. I've renamed it already, there's now `get_pages` (get the raw JSON of the pages), `get_item_collection` (gets pages as item collections), `get_items` (iterator through all pages and items), and `get_all_items` (gets all items from all pages and returns as single ItemCollection).   This matches the `get_` syntax used in PySTAC